File: sdks/python/joystauth.py
import os
import sys
import json
import re
import subprocess
import threading
import time
import requests

class UserData:
    def __init__(self, username="", subscription="", expiry="", hwid="", ip=""):
        self.username = username
        self.subscription = subscription
        self.expiry = expiry
        self.expires = expiry
        self.hwid = hwid
        self.ip = ip

class ResponseData:
    def __init__(self, success=False, message="", is_maintenance=False, active_notification=""):
        self.success = success
        self.message = message
        self.is_maintenance = is_maintenance
        self.active_notification = active_notification

class JoystAuth:
    """
    Joyst Corporation Auth Python SDK (100% Zero-Boilerplate Autopilot)
    """
    def __init__(self, name: str, token: str, version: str = "1.0", url: str = "https://joystauth.cc"):
        self.name = name
        self.token = token
        self.version = version
        self.url = (url or "https://joystauth.cc").rstrip("/")
        
        self.sessionid = None
        self.is_initialized = False
        self.hwid = self._get_hwid()
        self.user_data = UserData()
        self.response = ResponseData()
        self._last_notification = ""

        # ⚡ 1. Auto-init on launch
        self.init(auto_exit_on_maint=True)

        # ⚡ 2. Auto-launch background real-time heartbeat watchdog
        self._start_heartbeat_watchdog()

    def _get_hwid(self) -> str:
        hwid_str = ""
        try:
            if sys.platform == "win32":
                # Get Windows User Account SID (KeyAuth format S-1-5-21-...)
                out = subprocess.check_output("whoami /user", shell=True, stderr=subprocess.DEVNULL).decode().strip()
                m = re.search(r'S-1-5-21-\d+-\d+-\d+-\d+', out)
                if m:
                    return m.group(0)
                # Fallback: wmic csproduct get uuid
                output = subprocess.check_output("wmic csproduct get uuid", shell=True, stderr=subprocess.DEVNULL).decode().strip()
                lines = [l.strip() for l in output.split("\n") if l.strip() and "UUID" not in l]
                if lines: return lines[0]
            if not hwid_str:
                import uuid
                hwid_str = str(uuid.getnode())
        except Exception:
            import uuid
            hwid_str = str(uuid.getnode())
        return hwid_str

    def init(self, auto_exit_on_maint: bool = True) -> bool:
        try:
            payload = {"app_name": self.name, "app_token": self.token, "version": self.version, "hwid": self.hwid}
            res = requests.post(f"{self.url}/api/v1/client/init", json=payload, timeout=8)
            data = res.json()

            if data.get("success"):
                self.sessionid = data.get("sessionid")
                self.is_initialized = True
                self.response.success = True
                self.response.message = data.get("message", "Initialized successfully")
                self.response.is_maintenance = False

                notifs = data.get("notifications", [])
                if notifs and auto_exit_on_maint:
                    n = notifs[0]
                    self._last_notification = f"{n.get('title', 'ANNOUNCEMENT')}\n{n.get('message', '')}"
                    print(f"\n📢 [JOYST NOTIFICATION] {n.get('title')}: {n.get('message')}\n")
                return True
            else:
                self.response.success = False
                self.response.message = data.get("message") or data.get("detail") or "Failed to connect to authentication server."
                self.response.is_maintenance = bool(data.get("is_maintenance"))

                if auto_exit_on_maint:
                    print(f"\n🚨 [JOYST ALERT] {self.response.message}\n")
                    sys.exit(1)
                return False
        except Exception as e:
            self.response.success = False
            self.response.message = f"Connection error: {str(e)}"
            return False

    def _start_heartbeat_watchdog(self):
        def _loop():
            while True:
                time.sleep(15)
                if not self.sessionid: continue
                try:
                    payload = {
                        "app_name": self.name,
                        "app_token": self.token,
                        "hwid": self.hwid,
                        "username": self.user_data.username,
                        "sessionid": self.sessionid
                    }
                    res = requests.post(f"{self.url}/api/v1/client/check", json=payload, timeout=5)
                    if res.status_code == 200:
                        d = res.json()
                        if not d.get("success"):
                            msg = d.get("message", "Application placed into maintenance mode.")
                            print(f"\n🚨 [JOYST SECURITY ALERT] {msg}\n")
                            os._exit(1)
                except Exception:
                    pass
        t = threading.Thread(target=_loop, daemon=True)
        t.start()

    def login(self, username: str, password: str) -> bool:
        try:
            payload = {
                "app_name": self.name,
                "app_token": self.token,
                "username": username.strip(),
                "password": password,
                "hwid": self.hwid,
                "sessionid": self.sessionid
            }
            res = requests.post(f"{self.url}/api/v1/client/login", json=payload, timeout=8)
            data = res.json()

            if data.get("success"):
                self.user_data.username = username
                self.user_data.subscription = data.get("subscription", "default")
                self.user_data.expiry = data.get("expires_at", "Lifetime")
                self.user_data.hwid = self.hwid
                self.response.success = True
                self.response.message = data.get("message", "Logged in successfully")
                return True
            else:
                self.response.success = False
                self.response.message = data.get("message", "Login failed")
                return False
        except Exception as e:
            self.response.success = False
            self.response.message = str(e)
            return False

    def license(self, key: str) -> bool:
        try:
            payload = {
                "app_name": self.name,
                "app_token": self.token,
                "license_key": key.strip(),
                "hwid": self.hwid,
                "sessionid": self.sessionid
            }
            res = requests.post(f"{self.url}/api/v1/client/license", json=payload, timeout=8)
            data = res.json()

            if data.get("success"):
                self.user_data.username = data.get("username", key)
                self.user_data.subscription = data.get("subscription", "VIP Tier")
                self.user_data.expiry = data.get("expires_at", "Lifetime")
                self.user_data.hwid = self.hwid
                self.response.success = True
                self.response.message = data.get("message", "License verified successfully")
                return True
            else:
                self.response.success = False
                self.response.message = data.get("message", "Invalid license key")
                return False
        except Exception as e:
            self.response.success = False
            self.response.message = str(e)
            return False

    def register(self, username: str, password: str, license_key: str) -> bool:
        try:
            payload = {
                "app_name": self.name,
                "app_token": self.token,
                "username": username.strip(),
                "password": password,
                "license_key": license_key.strip(),
                "hwid": self.hwid,
                "sessionid": self.sessionid
            }
            res = requests.post(f"{self.url}/api/v1/client/register", json=payload, timeout=8)
            data = res.json()

            if data.get("success"):
                self.user_data.username = username
                self.response.success = True
                self.response.message = data.get("message", "Registered successfully")
                return True
            else:
                self.response.success = False
                self.response.message = data.get("message", "Registration failed")
                return False
        except Exception as e:
            self.response.success = False
            self.response.message = str(e)
            return False

    def var(self, var_name: str) -> str:
        try:
            payload = {
                "app_name": self.name,
                "app_token": self.token,
                "var_name": var_name,
                "sessionid": self.sessionid
            }
            res = requests.post(f"{self.url}/api/v1/client/var", json=payload, timeout=8)
            data = res.json()
            return data.get("value", "")
        except Exception:
            return ""

File: sdks/csharp/JoystAuth.cs
using System;
using System.IO;
using System.Net.Http;
using System.Security.Cryptography;
using System.Security.Principal;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using System.Diagnostics;
using Microsoft.Win32;

namespace JoystAuth
{
    public class SecurityShield
    {
        public static bool CheckCheatEngineInstalled()
        {
            try
            {
                // 1. Process Check
                string[] ceProcs = { "cheatengine", "cheatengine-x86_64", "cheatengine-i386", "processhacker", "x64dbg", "x32dbg", "ida64", "dnspy" };
                foreach (var p in Process.GetProcesses())
                {
                    string pName = p.ProcessName.ToLower();
                    foreach (var cep in ceProcs)
                    {
                        if (pName.Contains(cep)) return true;
                    }
                }

                // 2. Registry Check
                string[] regKeys = { @"Software\Cheat Engine", @"SOFTWARE\Cheat Engine", @"SOFTWARE\WOW6432Node\Cheat Engine" };
                foreach (var rk in regKeys)
                {
                    using var k1 = Registry.CurrentUser.OpenSubKey(rk);
                    if (k1 != null) return true;
                    using var k2 = Registry.LocalMachine.OpenSubKey(rk);
                    if (k2 != null) return true;
                }

                // 3. Directory Check
                string[] dirs = {
                    @"C:\Program Files\Cheat Engine",
                    @"C:\Program Files (x86)\Cheat Engine",
                    Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "Cheat Engine")
                };
                foreach (var d in dirs)
                {
                    if (Directory.Exists(d)) return true;
                }
            }
            catch { }
            return false;
        }
    }

    public class UserData
    {
        public string username { get; set; } = "";
        public string subscription { get; set; } = "default";
        public string expiry { get; set; } = "Lifetime";
        public string hwid { get; set; } = "";
        public string ip { get; set; } = "";
    }

    public class ResponseData
    {
        public bool success { get; set; } = false;
        public string message { get; set; } = "";
    }

    public class api
    {
        public string name { get; }
        public string token { get; }
        public string version { get; }
        public string url { get; }
        public string hwid { get; }
        public string sessionid { get; private set; } = "";

        public UserData user_data { get; } = new UserData();
        public ResponseData response { get; } = new ResponseData();

        private static readonly HttpClient client = new HttpClient { Timeout = TimeSpan.FromSeconds(8) };

        public api(string name, string token, string version = "1.0", string url = "https://joystauth.cc")
        {
            this.name = name;
            this.token = token;
            this.version = version;
            this.url = (url ?? "https://joystauth.cc").TrimEnd('/');
            this.hwid = GetHWID();

            if (SecurityShield.CheckCheatEngineInstalled())
            {
                Environment.Exit(0);
            }

            // ⚡ 1. Inbuilt Auto-Init
            this.init(true);

            // ⚡ 2. Inbuilt Live Heartbeat Watchdog
            this.StartHeartbeatWatchdog();
        }

        // Exact Windows User Account SID (KeyAuth format S-1-5-21-...)
        private string GetHWID()
        {
            try
            {
                var identity = WindowsIdentity.GetCurrent();
                if (identity?.User != null)
                {
                    return identity.User.Value;
                }
            }
            catch { }

            try
            {
                return Environment.MachineName + "_" + Environment.UserName;
            }
            catch { return "UNKNOWN_HWID"; }
        }

        public bool init(bool autoExitOnMaint = true)
        {
            try
            {
                if (SecurityShield.CheckCheatEngineInstalled()) Environment.Exit(0);

                var payload = new { app_name = this.name, app_token = this.token, version = this.version, hwid = this.hwid };
                var json = JsonSerializer.Serialize(payload);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var res = client.PostAsync($"{this.url}/api/v1/client/init", content).GetAwaiter().GetResult();
                var resStr = res.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                using var doc = JsonDocument.Parse(resStr);
                var root = doc.RootElement;

                bool ok = root.TryGetProperty("success", out var s) && s.GetBoolean();
                if (ok)
                {
                    this.sessionid = root.TryGetProperty("sessionid", out var sid) ? sid.GetString() : "";
                    this.response.success = true;
                    this.response.message = root.TryGetProperty("message", out var msg) ? msg.GetString() : "Initialized";
                    return true;
                }
                else
                {
                    this.response.success = false;
                    string msg = root.TryGetProperty("message", out var m) ? m.GetString() : (root.TryGetProperty("detail", out var d) ? d.GetString() : "Connection failed");
                    this.response.message = msg;

                    bool isMaint = (root.TryGetProperty("is_maintenance", out var im) && im.GetBoolean()) || msg.ToLower().Contains("maintenance");
                    if (autoExitOnMaint && isMaint)
                    {
                        Environment.Exit(0);
                    }
                    return false;
                }
            }
            catch (Exception ex)
            {
                this.response.success = false;
                this.response.message = ex.Message;
                return false;
            }
        }

        public bool login(string username, string password)
        {
            try
            {
                if (SecurityShield.CheckCheatEngineInstalled()) Environment.Exit(0);

                var payload = new { app_name = this.name, app_token = this.token, username = username.Trim(), password = password, hwid = this.hwid, sessionid = this.sessionid };
                var json = JsonSerializer.Serialize(payload);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var res = client.PostAsync($"{this.url}/api/v1/client/login", content).GetAwaiter().GetResult();
                var resStr = res.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                using var doc = JsonDocument.Parse(resStr);
                var root = doc.RootElement;

                bool ok = root.TryGetProperty("success", out var s) && s.GetBoolean();
                if (ok)
                {
                    this.user_data.username = username;
                    this.user_data.subscription = root.TryGetProperty("subscription", out var sub) ? sub.GetString() : "default";
                    this.user_data.expiry = root.TryGetProperty("expires_at", out var exp) ? exp.GetString() : "Lifetime";
                    this.user_data.hwid = this.hwid;
                    this.response.success = true;
                    this.response.message = root.TryGetProperty("message", out var msg) ? msg.GetString() : "Logged in";
                    return true;
                }
                else
                {
                    this.response.success = false;
                    this.response.message = root.TryGetProperty("message", out var msg) ? msg.GetString() : "Login failed";
                    return false;
                }
            }
            catch (Exception ex)
            {
                this.response.success = false;
                this.response.message = ex.Message;
                return false;
            }
        }

        public bool license(string key)
        {
            try
            {
                if (SecurityShield.CheckCheatEngineInstalled()) Environment.Exit(0);

                var payload = new { app_name = this.name, app_token = this.token, license_key = key.Trim(), hwid = this.hwid, sessionid = this.sessionid };
                var json = JsonSerializer.Serialize(payload);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var res = client.PostAsync($"{this.url}/api/v1/client/license", content).GetAwaiter().GetResult();
                var resStr = res.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                using var doc = JsonDocument.Parse(resStr);
                var root = doc.RootElement;

                bool ok = root.TryGetProperty("success", out var s) && s.GetBoolean();
                if (ok)
                {
                    this.user_data.username = root.TryGetProperty("username", out var u) ? u.GetString() : key;
                    this.user_data.subscription = root.TryGetProperty("subscription", out var sub) ? sub.GetString() : "VIP Tier";
                    this.user_data.expiry = root.TryGetProperty("expires_at", out var exp) ? exp.GetString() : "Lifetime";
                    this.user_data.hwid = this.hwid;
                    this.response.success = true;
                    this.response.message = root.TryGetProperty("message", out var msg) ? msg.GetString() : "License Active";
                    return true;
                }
                else
                {
                    this.response.success = false;
                    this.response.message = root.TryGetProperty("message", out var msg) ? msg.GetString() : "Invalid License";
                    return false;
                }
            }
            catch (Exception ex)
            {
                this.response.success = false;
                this.response.message = ex.Message;
                return false;
            }
        }

        public bool register(string username, string password, string licenseKey)
        {
            try
            {
                if (SecurityShield.CheckCheatEngineInstalled()) Environment.Exit(0);

                var payload = new { app_name = this.name, app_token = this.token, username = username.Trim(), password = password, license_key = licenseKey.Trim(), hwid = this.hwid, sessionid = this.sessionid };
                var json = JsonSerializer.Serialize(payload);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var res = client.PostAsync($"{this.url}/api/v1/client/register", content).GetAwaiter().GetResult();
                var resStr = res.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                using var doc = JsonDocument.Parse(resStr);
                var root = doc.RootElement;

                bool ok = root.TryGetProperty("success", out var s) && s.GetBoolean();
                if (ok)
                {
                    this.user_data.username = username;
                    this.response.success = true;
                    this.response.message = root.TryGetProperty("message", out var msg) ? msg.GetString() : "Registered successfully";
                    return true;
                }
                else
                {
                    this.response.success = false;
                    this.response.message = root.TryGetProperty("message", out var msg) ? msg.GetString() : "Registration failed";
                    return false;
                }
            }
            catch (Exception ex)
            {
                this.response.success = false;
                this.response.message = ex.Message;
                return false;
            }
        }

        private void StartHeartbeatWatchdog()
        {
            Task.Run(async () =>
            {
                while (true)
                {
                    await Task.Delay(25000);
                    if (string.IsNullOrEmpty(this.sessionid)) continue;
                    if (SecurityShield.CheckCheatEngineInstalled()) Environment.Exit(0);

                    try
                    {
                        var payload = new { app_name = this.name, app_token = this.token, sessionid = this.sessionid, hwid = this.hwid };
                        var json = JsonSerializer.Serialize(payload);
                        var content = new StringContent(json, Encoding.UTF8, "application/json");
                        var res = await client.PostAsync($"{this.url}/api/v1/client/check", content);
                        var resStr = await res.Content.ReadAsStringAsync();
                        using var doc = JsonDocument.Parse(resStr);
                        var root = doc.RootElement;

                        if (root.TryGetProperty("is_maintenance", out var im) && im.GetBoolean())
                        {
                            Environment.Exit(0);
                        }

                        if (root.TryGetProperty("success", out var ok) && !ok.GetBoolean())
                        {
                            Environment.Exit(0);
                        }
                    }
                    catch { }
                }
            });
        }
    }
}
